Add SubView class

gberg617 opened the pull request:
# Summary

- This WIP PR adds the SubView class.  This class stores a view-type object and a tuple of slices that indicate how to access the view's underlying data.  The following slice types are supported:
  - range slice: Indexes a subset of the original dimension's extents.  Does not reduce dimensionality.
  - fixed slice: Indexes along a fixed value of a given dimension.  Reduces dimensionality.
  - no slice: Indexes according to the original view's layout along the given dimension.  Does not reduce dimensionality.

A few simple examples are provided below:
```
    View<Index_type, Layout<2>> view(&my_data[0][0], Layout<2>(3,3));

    // "sv1 = View[1:3,:]"
    auto sv1 = SubView(view, RangeSlice{1,3}, NoSlice{});

    // "sv2 = View[1:3,1]"
    auto sv2 = SubView(view, RangeSlice{1,3}, FixedSlice{1});
```

Here, `sv1` and `sv2` are subviews of dimensions `(2,3)` and `(2)` respectively.

Future tasks:

- Add Slices whose extents are determined at compile time
- Add Slice types that use non-affine mappings (e.g. index lists)
- Add ability to collapse nested sub-layouts into a single sub-layout for slices that have affine mappings
- Add ability to distinguish between projected dimension and an empty slice (e.g. RangeSlice{2,2})

gberg617 committed:
first commit. WIP.

gberg617 committed:
Added tests

gberg617 committed:
small fixes

gberg617 committed:
updates to test

gberg617 commented:
This PR is still a WIP!

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,96 @@
+{
+
+// Slice descriptors
+struct RangeSlice { long start, end; };
+struct FixedSlice { long idx; };
+struct NoSlice { };
```
We will probably want to generalize the integer types used in the slices. We will probably want to add strongly typed slices like the strongly typed indices.

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,96 @@
+}
+
+template <typename T, size_t N, RAJA::Index_type... Is>
+RAJA_INLINE RAJA_HOST_DEVICE constexpr auto array_to_tuple_impl(const camp::array<T, N>& arr, camp::idx_seq<Is...>) {
+    return camp::make_tuple(arr[Is]...);
+}
+
+template <typename T, size_t N>
+RAJA_INLINE RAJA_HOST_DEVICE constexpr auto array_to_tuple(const camp::array<T, N>& arr) {
+    return array_to_tuple_impl(arr, camp::make_idx_seq_t<N>{});
+}
```
Do we not have this somewhere?

gberg617 replied:
This function wasn't eventually needed.  That said, I can still add something like this to camp if you think it could be useful.

gberg617 committed:
various changes/fixes to address comments

gberg617 commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,96 @@
+                    } else {
+                        return idxs[idx++];
+                    }
+                }()
```
I moved it to a member function.

gberg617 commented on test/unit/view-layout/test-subview.cpp:
```diff
@@ -0,0 +1,120 @@
+
+}
+
+// void test_subviewGPU() {
```
To Do: Fix GPU tests.

gberg617 replied:
Perhaps remove GPU tests?

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,120 @@
+struct NoSlice { 
+    static constexpr bool reduces_dimension = false;
+
+    template<typename IndexType = Index_type>
```
Would it make more sense to template all of these classes on IndexType so they are consistent?

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,120 @@
+// Helper to count non-fixed dimensions at compile time
+template <typename... Slices>
+RAJA_INLINE RAJA_HOST_DEVICE constexpr size_t count_nonfixed_dims() {
+    return (!Slices::reduces_dimension + ...);
```
Are we adding booleans here? How about a ternary with 1 and 0?

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,120 @@
+class SubView {
+    ViewType view_;
+    camp::tuple<Slices...> slices_;
+    std::array<IndexType, sizeof...(Slices)> map_;
+
+    RAJA_INLINE RAJA_HOST_DEVICE constexpr void make_subview_index_map() {
+        size_t sub_idx = 0;
+        size_t i = 0;
+        ((map_[i++] = (Slices::reduces_dimension ? -1 : sub_idx++)), ...);
+    }
```
It would probably be more clear if make_subview_index_map returned an array. If possible it would be nice to generate this map in a way that looks as much like a normal for loop doing a scan as possible. I think there is a RAJA util like for_each_type that might be useful here.

gberg617 replied:
Fixed.

gberg617 replied:
I found having a `for_each` that calls a func with the tuple and index seemed to be the cleanest.  Let me know your thoughts.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,120 @@
+    return array_to_tuple_impl(arr, camp::make_idx_seq_t<N>{});
+}
+
+template <typename ViewType, typename IndexType = Index_type, typename... Slices>
```
Think about doing using a camp::list<slices...> to avoid having a parameter pack in the class template.

gberg617 replied:
I tried doing something like that.  I'm not quite sure I had the right approach.

rcarson3 commented on test/unit/view-layout/test-subview.cpp:
```diff
@@ -0,0 +1,120 @@
+    EXPECT_EQ(sv(1,2), 9);
+
+}
+
```
@gberg617 how would you do something like we have in the original impl of SubViews: https://github.com/LLNL/SNLS/blob/develop/test/SNLS_forall_subviews.cxx#L69-L75 where we allow for a sliding window with the SubView. Additionally, how can we get the underlying data pointer at the SubView's (0..) index like here: https://github.com/LLNL/SNLS/blob/develop/test/SNLS_forall_subviews.cxx#L150 ?

gberg617 replied:
Added `set_slice` to allow things like "sliding windows"

gberg617 committed:
implemented for_each_index

gberg617 committed:
various more simplifications

gberg617 committed:
added set/get slice methods and test

gberg617 commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,120 @@
+
+    template <typename... Idxs>
+    RAJA_INLINE RAJA_HOST_DEVICE constexpr IndexType operator()(Idxs... idxs) const {
+        constexpr size_t nidx = count_nonfixed_dims<Slices...>();
```
I simplified this a bit

gberg617 commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,120 @@
+
+    template<IndexType I>
+    RAJA_INLINE RAJA_HOST_DEVICE constexpr auto map_subview_idx_to_parent(IndexType* idxs) const {
+        return camp::get<I>(slices_).map_index(idxs[map_[I]]);
```
I simplified this a bit.  Let me know your thoughts.

gberg617 commented:
I'm considering separating the SubView holding the view from the slices that are used to access the view,  somewhat analogous to the approach of having layouts and Views be separate classes.  Any thoughts on this?

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,144 @@
+class SubView<ViewType, camp::list<Slices...>, IndexType> {
+    ViewType view_;
+    camp::tuple<Slices...> slices_;
+    std::array<IndexType, sizeof...(Slices)> map_;
```
Still need to make this static inline constexpr

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,144 @@
+        size_t sub_idx = 0;
+        std::array<IndexType, sizeof...(Slices)> map;
+
+        for_each_tuple_index( slices_, 
```
This looks great. Unfortunately it uses a non-static member to do something static.

gberg617 replied:
I think this might be outdated?

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,144 @@
+
+    template <typename... Idxs>
+    RAJA_INLINE RAJA_HOST_DEVICE constexpr IndexType operator()(Idxs... idxs) const {
+        constexpr size_t nidx = ((Slices::reduces_dimension == false ? 1 : 0) + ...);
```
This should be a static member variable of the class.

MrBurmark replied:
It would be nice to have something like num_slices as well.

MrBurmark replied:
It would also be nice to have a bit more descriptive names.

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,144 @@
+
+        for_each_tuple_index( slices_,
+            [&](auto slice, auto index) { 
+                parent_indices[index] = slice.map_index(arr[map_[index]]); 
```
This causes out of bounds accesses for slices that reduce dimensions.

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,144 @@
+                parent_indices[index] = slice.map_index(arr[map_[index]]); 
+            });
+
+        return camp::apply(view_, array_to_tuple(parent_indices));
```
Why not just use a tuple to start with?

gberg617 replied:
I think this is also outdated.  We moved to using camp::array.

MrBurmark commented on include/RAJA/util/for_each.hpp:
```diff
@@ -79,8 +96,31 @@ RAJA_HOST_DEVICE RAJA_INLINE UnaryFunc for_each_tuple(Tuple&& t,
+{
+  using camp::get;
+  // braced init lists are evaluated in order
+  int seq_unused_array[] = {0, (func(get<Is>(std::forward<Tuple>(t)), Is), 0)...};
```
```suggestion
  int seq_unused_array[] = {0, (func(get<Is>(std::forward<Tuple>(t)), std::integral_constant<std::size_t, Is>{}), 0)...};
```

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/for_each.hpp:
```diff
@@ -36,6 +39,20 @@ namespace RAJA
+// compile time expansion applying func to each of the indices
+RAJA_SUPPRESS_HD_WARN
+template<typename UnaryFunc, typename IndexType, IndexType... Is>
+RAJA_HOST_DEVICE RAJA_INLINE UnaryFunc for_each_index(camp::idx_seq<Is...>,
```
```suggestion
template<typename UnaryFunc, camp::idx_t... Is>
RAJA_HOST_DEVICE RAJA_INLINE UnaryFunc for_each_index(camp::idx_seq<Is...>,
```

MrBurmark replied:
It should all be size_t or camp::idx_t.

gberg617 replied:
Fixed.

MrBurmark commented:
> I'm considering separating the SubView holding the view from the slices that are used to access the view, somewhat analogous to the approach of having layouts and Views be separate classes. Any thoughts on this?

You mean something more like making a SubLayout that you can use anywhere a layout can be used?

gberg617 commented:
> > I'm considering separating the SubView holding the view from the slices that are used to access the view, somewhat analogous to the approach of having layouts and Views be separate classes. Any thoughts on this?
> 
> You mean something more like making a SubLayout that you can use anywhere a layout can be used?

Yes, but the intent is that these "sub-layouts" are only used with SubViews.  I will try to prototype something here soon.

gberg617 committed:
made simplifications to map and parent_indices

gberg617 committed:
made simplifications to map and parent_indices

gberg617 committed:
started adding various size/dim methods

gberg617 committed:
started implementing SubLayout

MrBurmark commented:
> > You mean something more like making a SubLayout that you can use anywhere a layout can be used?
> 
> Yes, but the intent is that these "sub-layouts" are only used with SubViews. I will try to prototype something here soon.

We'd have to think about it. It seems simpler to me if you can use a sublayout like a layout, that way a subview would just be a view using a sublayout.

gberg617 committed:
first attempt at changing to SubLayout

gberg617 committed:
made some changes to testing subview/sublayouts

gberg617 commented on include/RAJA/util/TypedViewBase.hpp:
```diff
@@ -669,16 +669,31 @@ class ViewBase
   RAJA_INLINE
   constexpr linear_index_type size() const { return m_layout.size(); }
 
+  RAJA_HOST_DEVICE
```
changes to View and MultiView are still a WIP.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+    template<IndexType DIM, typename LayoutType>
+    RAJA_INLINE RAJA_HOST_DEVICE constexpr IndexType size(const LayoutType&) const {
+        return (end_ - start_ + 1);
+    }
```
Why is DIM here? If its unused we should use the unused arg macro so we don't get a warning.

MrBurmark replied:
Also why the +1, this is including end?

gberg617 replied:
RangeSlice includes the start and end elements, which is why we need the +1.  So `RangeSlice<>{1,3}` will have a size of 3.  Would it make more sense to have the range be `[start, end)`? Perhaps that's the more standard approach.

rhornung67 replied:
We've discussed this and I thought we had an issue for it. But I can't find it. Let's discuss at the project meeting tomorrow.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+
+template<typename IndexType = Index_type>
+struct StridedSlice { 
+    IndexType start_, end_, stride_;
```
Does this support negative strides? I'm pretty sure that the strided range supports negative strides.

rhornung67 replied:
It does support negative strides. For example, specifying stride = -1 is how we reverse the iteration order of a loop.

gberg617 replied:
I added negative stride support in size().  It's implicitly assumed here that start/end ordering are consistent with stride sign.  So start < end when stride > 0, and start > end when stride < 0.  I can add run-time checks for this if needed.

gberg617 replied:
I ended up adding more checks to size() for StridedSlice.  Would it make sense at all to generally consider supporting negative indexing as well?

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+RAJA_INLINE RAJA_HOST_DEVICE constexpr auto make_slice_to_parent_index_map() {
+    IndexType sub_idx = 0;
+    IndexType i = 0;
+    camp::array<IndexType, sizeof...(Slices)> map = {};
```
Why not do something like this?
```suggestion
    camp::array<IndexType, sizeof...(Slices)> map{};
```

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+    IndexType sub_idx = 0;
+    IndexType i = 0;
+    camp::array<IndexType, sizeof...(Slices)> map = {};
+    ((map[i++] = (Slices::reduces_dimension ? -1 : sub_idx++)), ...);
```
Or even do something like this?
```suggestion
    camp::array<IndexType, sizeof...(Slices)> map{{(Slices::reduces_dimension ? -1 : sub_idx++)...}};
```

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+}
+
+template<typename IndexType, size_t n_parent_dims, typename... Slices>
+RAJA_INLINE RAJA_HOST_DEVICE constexpr auto make_parent_to_slice_index_map() {
```
Does it make more sense to take n_parent_dims as an argument or calculate it inside?

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+
+    using IndexLinear = IndexType;
+
+    const LayoutType& parent_;
```
Instead of having a copy of the parent this has a reference? What if you make the SubRegion on the host and then copy it to the device?

MrBurmark replied:
Also what do you think about using the member variables are named with m_ and static members are named with s_ convention?

gberg617 replied:
Fixed, except for `n_dims` which had to stay as it was.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+    const LayoutType& parent_;
+    camp::tuple<Slices...> slices_;
+
+    static inline constexpr size_t num_slices_ = sizeof...(Slices);
```
Static data members and functions should be above non-static data members and functions.

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+    static inline constexpr size_t num_slices_ = sizeof...(Slices);
+
+    static inline constexpr 
+    IndexType n_dims = ((Slices::reduces_dimension == false ? 1 : 0) + ...);
```
How about using `Slices::reduces_dimension` as a boolean if its a boolean?

MrBurmark replied:
I like to have all the uses of something look similar if possible. I think this is often used like this `!Slices::reduces_dimension` so it would be simpler to understand if this also looked the same.

gberg617 replied:
Do you mean computing n_dims like this: `n_dims = (!Slices::reduces_dimension + ...)`?

MrBurmark replied:
no I mean
```
((!Slices::reduces_dimension ? 1 : 0) + ...)
```

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+    template<IndexType Index>
+    RAJA_INLINE RAJA_HOST_DEVICE constexpr auto& get_slice() const {
+        return camp::get<Index>(slices_);
+    }
```
Since there are const they should probably return const references.

gberg617 replied:
Fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+        for_each_tuple_index( slices_,
+            [&](auto slice, auto index) {
+                const IndexType dim_size = slice.template size<index>(parent_);
+                prod_dims *= (dim_size == 0) ? 1 : dim_size;
```
If there really is a dimension that is completely projected out, shouldn't the size be 0?

MrBurmark replied:
Also do we have to consider the 0 dimension case?

gberg617 replied:
I'm not sure what you mean here.  The intent was to keep the behavior consistent with what's in Layout: size() will replace zero-sized dimensions with 1 when computing the product of sizes.

MrBurmark replied:
Didn't we agreed that Layout is wrong and that if a dimension has zero items then the product should be 0. I guess its ok to be consistent but I thought we were going to fix that in Layout in the next major release.

MrBurmark replied:
I suppose that completely projected out just means that a single value was chosen? So the size is one in that direction?

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+    template<IndexType DIM> 
+    RAJA_INLINE RAJA_HOST_DEVICE constexpr auto get_dim_size() const {
+        constexpr auto SliceDim = parent_to_slice_map_[DIM];
+        return camp::get<SliceDim>(slices_).template size<DIM>(parent_);
```
We should be able to static_assert that DIM is in bounds.

gberg617 replied:
Added.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+
+        for_each_tuple_index( slices_,
+            [&](auto slice, auto index) {
+                if (slice_to_parent_map_[index] >= 0) {
```
should this be if constexpr?

gberg617 replied:
Yes, fixed.

MrBurmark commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,268 @@
+                    // map_index will not need index values for dimension-reducing slices
+                    // so we pass a "dummy" value.
+                    constexpr IndexType dummy_value = -1;
+                    parent_indices[index] = slice.map_index(dummy_value); 
```
Should dimension reducing slices not take a value at all?

gberg617 replied:
Yup.  I added a new map_index() function that takes no arguments for dimension reducing slices.

gberg617 committed:
Made various fixes to address comments.

gberg617 committed:
added negative stride support for StridedSlice

gberg617 commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,272 @@
+
+    template<IndexType RAJA_UNUSED_ARG(DIM), typename LayoutType>
+    RAJA_INLINE RAJA_HOST_DEVICE constexpr IndexType size(const LayoutType&) const {
+        return (end_ - start_ + 1);
```
Should we consider changing RangeSlice to the half-open interval `[start_, end_)` instead of the closed interval `[start_, end_]`?

MrBurmark replied:
That would be more consistent with how most things work in C++.

gberg617 replied:
Fixed.

gberg617 committed:
modified RangeSlice and StridedSlice sizes to only include the half-open interval [start, end) and added more checks for StridedSlice size

artv3 commented on tpl/camp:
does camp need to be updated?

gberg617 replied:
Yes, it does.  Thanks for catching that.  See https://github.com/llnl/camp/pull/194.

gberg617 committed:
minor fixes to address comments

gberg617 committed:
added n_dims to all view types and added n_dims == num_slices static_assert to SubRegion

gberg617 commented on include/RAJA/util/View.hpp:
```diff
@@ -165,6 +165,8 @@ struct MultiView
   using NonConstView =
       MultiView<nc_value_type, layout_type, P2Pidx, nc_pointer_type>;
 
+  static constexpr size_t n_dims = layout_type::n_dims + 1;
```
@rchen20 Does this look right for MultiViews?

gberg617 commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,120 @@
+template <typename ViewType, typename IndexType = Index_type, typename... Slices>
+class SubView {
+    ViewType view_;
+    camp::tuple<Slices...> slices_;
```
This compile-time check is a good idea.  The best way I could think of doing this is adding `n_dims` to each view type to have a consistent way of checking for the number of dimensions in the view. See this commit for those changes: https://github.com/llnl/RAJA/pull/1915/changes/0b198afb679758a08e5856af92d85ba48baf638d.

gberg617 committed:
modifications to size(), size_noproj(), and get_dim_stride()

gberg617 committed:
Added draft version of get_dim_logical_stride and testing for sizes with projected dimensions

gberg617 committed:
1. Renamed start_,end_ slice members to m_start and m_end.
2. Replaced camp::array usage with std::array.
3. Fixed and simplified get_dim_stride and get_parent_dim_stride.
4. Replaced usage of IndexType with size_t where appropriate.

gberg617 commented on include/RAJA/util/SubView.hpp:
```diff
@@ -263,15 +243,28 @@ struct SubRegion<LayoutType, camp::list<Slices...>, IndexType> {
+    }
+
+    template<size_t DIM>
+    RAJA_INLINE RAJA_HOST_DEVICE constexpr auto get_parent_dim_stride() const {
```
I'm not sure how useful `get_parent_dim_stride()` is.

gberg617 committed:
added HIP subview test

gberg617 committed:
added HIP test and finished removing camp::array

gberg617 commented on include/RAJA/util/SubView.hpp:
```diff
@@ -0,0 +1,298 @@
+    RAJA_INLINE RAJA_HOST_DEVICE constexpr auto operator()(Idxs... idxs) const {
+        static_assert(sizeof...(idxs) == n_dims, "Wrong number of indices");
+
+        std::array<IndexType, n_dims> arr{idxs...};
```
@MrBurmark Did we decide to go with `std::array`, or switch back to `camp::array` here?

MrBurmark replied:
Let's use camp array, we haven't asked users to add the cuda flag that would enable this yet.

gberg617 replied:
OK, I added back camp::array.  This means that we first need to push the following camp PR:
https://github.com/llnl/camp/pull/194

gberg617 committed:
apply clang format

gberg617 committed:
update submodules to state in develop

gberg617 committed:
empty commit

gberg617 committed:
added subview documentation using codex

gberg617 committed:
added back camp::array in place of std::array

gberg617 committed:
revert radiuss-spack-configs to commit used in develop

gberg617 committed:
update camp

gberg617 commented on include/RAJA/util/for_each.hpp:
```diff
@@ -26,6 +26,9 @@
 #include <type_traits>
 
```
I need to fix this rebased file.

rhornung67 commented on test/unit/view-layout/test-subview.cpp:
```diff
@@ -1,5 +1,5 @@
 //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
-// Copyright (c) 2016-25, Lawrence Livermore National Security, LLC
+// Copyright (c) 2016-26, Lawrence Livermore National Security, LLC
```
@gberg617 we changed the copyright header in all RAJA files so when we update each year we only need to touch two files rather than the whole repo. Here is what you should have on all files you add or modify: 
https://github.com/llnl/RAJA/blob/develop/test/unit/view-layout/test-indexlayout.cpp#L1

rhornung67 reviewed (changes requested):
See comment about copyright header.

gberg617 commented:
@MrBurmark @adayton1 I assume that SubViews can be 0D in certain cases where the slice(s) reduce the dimension to a scalar. This leads to compiler errors on Windows (see failing CI) because camp array stores a C-style array of size 0. My thoughts are to either disallow 0D SubViews or add a specialization of camp array for a 0-sized array. My preference is the latter because std::array can be zero-sized, so this addition would make camp::array be more consistent with std::array. Any thoughts?

MrBurmark commented:
> @MrBurmark @adayton1 I assume that SubViews can be 0D in certain cases where the slice(s) reduce the dimension to a scalar. This leads to compiler errors on Windows (see failing CI) because camp array stores a C-style array of size 0. My thoughts are to either disallow 0D SubViews or add a specialization of camp array for a 0-sized array. My preference is the latter because std::array can be zero-sized, so this addition would make camp::array be more consistent with std::array. Any thoughts?

Adding a camp array specialization for 0-size makes a lot of sense.

adayton1 commented:
> > @MrBurmark @adayton1 I assume that SubViews can be 0D in certain cases where the slice(s) reduce the dimension to a scalar. This leads to compiler errors on Windows (see failing CI) because camp array stores a C-style array of size 0. My thoughts are to either disallow 0D SubViews or add a specialization of camp array for a 0-sized array. My preference is the latter because std::array can be zero-sized, so this addition would make camp::array be more consistent with std::array. Any thoughts?
> 
> Adding a camp array specialization for 0-size makes a lot of sense.

I can't remember for sure, but I think that I disallowed the 0 size array because nvcc couldn't handle it (or maybe it was some other compiler). So just do plenty of testing.
